src/common/type/pack.c

```diff
@@ -0,0 +1,1571 @@
+
+    FUNCTION_TEST_RETURN_VOID();
+}
+
+void
+pckReadArrayEnd(PackRead *this)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_READ, this);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+
+    if (lstSize(this->tagStack) == 1 || this->tagStackTop->type != pckTypeArray)
+        THROW(FormatError, "not in array");
+
+    // Make sure we are at the end of the array
+    unsigned int id = 0xFFFFFFFF - 1;
+    pckReadTag(this, &id, pckTypeUnknown, true);
+
+    // Pop array off the stack
+    lstRemoveLast(this->tagStack);
+    this->tagStackTop = lstGetLast(this->tagStack);
+
+    // Reset tagNextId to keep reading
+    this->tagNextId = 0;
+
+    FUNCTION_TEST_RETURN_VOID();
+}
+
+/**********************************************************************************************************************************/
+Buffer *
+pckReadBin(PackRead *this, PckReadBinParam param)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_READ, this);
+        FUNCTION_TEST_PARAM(UINT, param.id);
+        FUNCTION_TEST_PARAM(BOOL, param.defaultNull);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+
+    if (pckReadDefaultNull(this, param.defaultNull, &param.id))
+        FUNCTION_TEST_RETURN(NULL);
+
+    Buffer *result = NULL;
+
+    if (pckReadTag(this, &param.id, pckTypeBin, false))
+    {
+        result = bufNew((size_t)pckReadUInt64Internal(this));
+
+        while (bufUsed(result) < bufSize(result))
+        {
+            size_t size = pckReadBuffer(this, bufRemains(result));
+            bufCatC(result, this->bufferPtr, this->bufferPos, size);
+            this->bufferPos += size;
+        }
+    }
+    else
+        result = bufNew(0);
+
+    FUNCTION_TEST_RETURN(result);
+}
+
+/**********************************************************************************************************************************/
+bool
+pckReadBool(PackRead *this, PckReadBoolParam param)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_READ, this);
+        FUNCTION_TEST_PARAM(UINT, param.id);
+        FUNCTION_TEST_PARAM(BOOL, param.defaultNull);
+        FUNCTION_TEST_PARAM(BOOL, param.defaultValue);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+
+    if (pckReadDefaultNull(this, param.defaultNull, &param.id))
+        FUNCTION_TEST_RETURN(param.defaultValue);
+
+    FUNCTION_TEST_RETURN(pckReadTag(this, &param.id, pckTypeBool, false));
+}
+
+/**********************************************************************************************************************************/
+int32_t
+pckReadI32(PackRead *this, PckReadInt32Param param)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_READ, this);
+        FUNCTION_TEST_PARAM(UINT, param.id);
+        FUNCTION_TEST_PARAM(BOOL, param.defaultNull);
+        FUNCTION_TEST_PARAM(INT, param.defaultValue);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+
+    if (pckReadDefaultNull(this, param.defaultNull, &param.id))
```

Comment:
param.id is not always set so in those cases we're always going to look at the object that we consider the next to be read, right? But I guess everything about the pack is you must read and write in order so say I have all int32s then if I think I'm reading ID 4 with a default of 5 but I didn't specify the ID so I'm actually reading ID 3 because the last thing I read was ID 2, then I could be checking the defaultNull for the wrong ID. But that is the chance we take, right?

Reply:
Correct -- if you are going to be skipping fields you really need an enum that specifies the field ids to be used on both sides. There will be cases like this.

However, our current use case (the one in this patch) does not create any gaps so constant IDs are not needed. But the serializer/deserializer need to be coded with the same logic.